sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -98,6 +108,100 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
+    // to fully qualified the column names in condition. Therefore, query should be rewritten so
+    // that each of the outputs of child joins are aliased, so there would be a projection
+    // with aliases between top level join and scanBuilderHolder (that has pushed child joins).
+    case node @ Join(
+      PhysicalOperation(
+        leftProjections,
+        Nil,
+        leftHolder @ ScanBuilderHolder(_, _, lBuilder: SupportsPushDownJoin)
+      ),
+      PhysicalOperation(
+        rightProjections,
+        Nil,
+        rightHolder @ ScanBuilderHolder(_, _, rBuilder: SupportsPushDownJoin)
+      ),
+      joinType,
+      condition,
+    _) if conf.dataSourceV2JoinPushdown &&
+      // TODO: I think projections will always be Seq[AttributeReference] because
+      // When
+      // SELECT tbl1.col+2, tbl2.* FROM tbl1 JOIN tlb2
+      // is executed, col is pruned down, but col + 2 will be projected on top of join.
+      leftProjections.forall(_.isInstanceOf[AttributeReference]) &&
+      rightProjections.forall(_.isInstanceOf[AttributeReference]) &&
+      lBuilder.isRightSideCompatibleForJoin(rBuilder) =>
+      val normalizedLeftProjections = DataSourceStrategy.normalizeExprs(
+        leftProjections,
+        leftHolder.output
+      ).asInstanceOf[Seq[AttributeReference]]
+      val leftRequiredSchema = fromAttributes(normalizedLeftProjections)
+
+      val normalizedRightProjections = DataSourceStrategy.normalizeExprs(
+        rightProjections,
+        rightHolder.output
+      ).asInstanceOf[Seq[AttributeReference]]
+      val rightRequiredSchema = fromAttributes(normalizedRightProjections)
+
+      val normalizedCondition = condition.map { e =>
+        DataSourceStrategy.normalizeExprs(
+          Seq(e),
+          leftHolder.output ++ rightHolder.output
+        ).head
+      }
+
+      val conditionWithJoinColumns = normalizedCondition.map { cond =>
+        cond.transformUp {
+          case a: AttributeReference =>
+            val isInLeftSide = leftProjections.filter(_.exprId == a.exprId).nonEmpty
+            JoinColumnReference(a, isInLeftSide)
+        }
+      }
+
+      val translatedCondition =
+        conditionWithJoinColumns.flatMap(DataSourceV2Strategy.translateFilterV2(_))
+      val translatedJoinType = DataSourceStrategy.translateJoinType(joinType)
+
+      if (translatedCondition.isDefined == condition.isDefined &&
+        translatedJoinType.isDefined &&
+        lBuilder.pushJoin(
+          rBuilder,
+          translatedJoinType.get,
+          translatedCondition.toJava,
+          leftRequiredSchema,
+          rightRequiredSchema
+        )) {
+        leftHolder.joinedRelations = leftHolder.joinedRelations :+ rightHolder.relation
+
+        val newSchema = leftHolder.builder.build().readSchema()
+        val newOutput = (leftProjections ++ rightProjections).asInstanceOf[Seq[AttributeReference]]
+          .zip(newSchema.fields)
```

Comment:
We should fail if the number of columns doesn't match between Spark and the third-party data source

Reply:
also check the data type.